src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -533,26 +533,50 @@ private void RenderEndTagHelpersScope()
                                                   _tagHelperContext.ScopeManagerEndMethodName);
         }
 
-        private void RenderWriteTagHelperMethodCall(TagHelperChunk chunk)
+        private void RenderTagHelperOutput(TagHelperChunk chunk)
         {
+            var tagHelperOutputAccessor =
+                $"{ExecutionContextVariableName}.{_tagHelperContext.ExecutionContextOutputPropertyName}";
+
+            if (ContainsChildContent(chunk.Children))
```

Comment:
> Why not unconditionally write this check? Or at least only skip this output of there are no Children at all. Very special-case-y at the moment.

We can only do this at compile time because we know if there's whitespace etc. that exist within the child content.



> After looking at aspnet/Mvc#3582 doesn't look like the intricacies of this method are tested. Can see this method does the Right Thing with self-closing elements, those containing straight text, those containing tag helpers, and not much else.

We verify it, no worries. Several self-closing examples in the code generation tests. Heck we have every tag form in the tests :smile: